### PR TITLE
Add DiscordGuildMessagePreBroadcastEvent

### PR DESCRIPTION
**Pull request by Majekdor**
This makes it possible to change the recipients, destination channel, and message content of a message being sent from Discord -> Minecraft. In my [PartyChat](https://github.com/Majekdor/PartyChat) plugin I have a feature that allows players to toggle `party-only`, which means they only see PartyChat messages, not normal chat. I had a user request the ability to hide Discord messages with that same feature, which was impossible using DiscordSRV without this PR.

### CODE REVIEW

File: src/main/java/github/scarsz/discordsrv/DiscordSRV.java [hunk 1]
```diff
@@ -1768,9 +1765,13 @@ public void broadcastMessageToMinecraftServer(String channel, Component message,
...
                 .findAny().orElse(null);
 
+        DiscordGuildMessagePreBroadcastEvent preBroadcastEvent = api.callEvent(new DiscordGuildMessagePreBroadcastEvent
+                (channel, message, PlayerUtil.getOnlinePlayers()));
+        message = preBroadcastEvent.getMessage();
+        channel = preBroadcastEvent.getChannel();
         if (chatHook == null || channel == null) {
-            if (channel != null && !channel.equalsIgnoreCase("global")) return; // don't send messages for non-global channels with no plugin hooks
-            MessageUtil.sendMessage(PlayerUtil.getOnlinePlayers(), message);
+            if (channel != null && !channel.equals("global")) return; // don't send messages for non-global channels with no plugin hooks
```
Comment (Scarsz) [inline src/main/java/github/scarsz/discordsrv/DiscordSRV.java:1773]: Why the `String#equalsIgnoreCase` to `String#equals`

Comment (Majekdor) [reply src/main/java/github/scarsz/discordsrv/DiscordSRV.java:1773]: Honestly, I don't remember doing that or why I did it... will switch it back.

File: src/main/java/github/scarsz/discordsrv/api/events/DiscordGuildMessagePreBroadcastEvent.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import java.util.List;
+
+/**
+ * <p>Called directly before a message from Discord is processed and broadcasted to the Minecraft server</p>
+ *
+ * <p>At the time this event is called, the message, channel, and recipients can still be changed.</p>
+ */
+public class DiscordGuildMessagePreBroadcastEvent extends Event {
+
+    @Getter @Setter private  String channel;
```
Comment (Dinty1) [inline src/main/java/github/scarsz/discordsrv/api/events/DiscordGuildMessagePreBroadcastEvent.java:39]: Extra space here should maybe be removed for consistency

**Review by Dinty1 — CHANGES_REQUESTED** [review]

File: src/main/java/github/scarsz/discordsrv/DiscordSRV.java [hunk 3]
```diff
@@ -1768,9 +1765,13 @@ public void broadcastMessageToMinecraftServer(String channel, Component message,
...
+        DiscordGuildMessagePreBroadcastEvent preBroadcastEvent = api.callEvent(new DiscordGuildMessagePreBroadcastEvent
+                (channel, message, PlayerUtil.getOnlinePlayers()));
+        message = preBroadcastEvent.getMessage();
+        channel = preBroadcastEvent.getChannel();
         if (chatHook == null || channel == null) {
             if (channel != null && !channel.equalsIgnoreCase("global")) return; // don't send messages for non-global channels with no plugin hooks
-            MessageUtil.sendMessage(PlayerUtil.getOnlinePlayers(), message);
+            MessageUtil.sendMessage(preBroadcastEvent.getRecipients(), message);
             PlayerUtil.notifyPlayersOfMentions(null, MessageUtil.toLegacy(message));
         } else {
```
Comment (Vankka) [inline src/main/java/github/scarsz/discordsrv/DiscordSRV.java:1776]: You should either not run the event when chat hooks are in use, or add documentation/throw a exception for trying to change recipients when chat hooks are being used.

File: src/main/java/github/scarsz/discordsrv/api/events/DiscordGuildMessagePreBroadcastEvent.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    @Getter @Setter private Component message;
+    @Getter @Setter private List<? extends CommandSender> recipients;
```
Comment (Vankka) [inline src/main/java/github/scarsz/discordsrv/api/events/DiscordGuildMessagePreBroadcastEvent.java:41]: Don't know that having a setter for this is necessary, a getter is enough as Lists provide methods for altering their contents

File: src/main/java/github/scarsz/discordsrv/DiscordSRV.java (hunk 3 above)
Comment (Majekdor) [reply src/main/java/github/scarsz/discordsrv/DiscordSRV.java:1776]: I hadn't thought of that. I'll move it so it doesn't affect any chat hooks.

File: src/main/java/github/scarsz/discordsrv/api/events/DiscordGuildMessagePreBroadcastEvent.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+import org.bukkit.command.CommandSender;
+
+import java.util.List;
+
+/**
+ * <p>Called directly before a message from Discord is processed and broadcasted to the Minecraft server</p>
+ *
+ * <p>At the time this event is called, the message, channel, and recipients can still be changed.</p>
+ */
+public class DiscordGuildMessagePreBroadcastEvent extends Event {
```
Comment (Vankka) [inline src/main/java/github/scarsz/discordsrv/api/events/DiscordGuildMessagePreBroadcastEvent.java:37]: As this now isn't triggered when chat hooks are being used, a comment in the class would be useful

**Review by Vankka — APPROVED** [review]

**Pull request merged**